### PR TITLE
[PoemBot] Basic Blocks + library

### PR DESCRIPTION
**Pull request by ajpal**
Similar to https://github.com/code-dot-org/code-dot-org/pull/40363, but for Poem Bot.
More blocks to come, but this seems like a good checkpoint for now.
Trying to build it all to work with Google Blockly from the start:
image

## Testing story
There's no good way to write unit tests for code in interpreted libraries. Also, since this is just a prototype, I imagine the functionality will change quite a bit over the next month or two. We should definitely have unit test coverage before HOC, but I don't think it's that important at this point.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/config/libraries/PoemBot.interpreted.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+var poem = {
+  title: undefined,
+  author: undefined,
+  lines: []
+};
+var isVisible = false;
+
+function randomWord() {
+  return "cat";
```
Comment (ajpal) [inline dashboard/config/libraries/PoemBot.interpreted.js]: TODO: this will return a random word from a list curated by Amy BW but the list is not ready yet, so this is just a placeholder

Comment (breville) [reply dashboard/config/libraries/PoemBot.interpreted.js]: random thought: it could be fun to have it return a random word from a small list?  This would also help make sure that nothing weird happens when it returns different values.

Comment (breville) [reply dashboard/config/libraries/PoemBot.interpreted.js]: To elaborate, I remember long ago seeing something similar where things seemed fine, but actually the random generator was being called quite often, not just at init time, and once a variety of random values were being returned, a piece of UI that was supposed to be stable was actually constantly changing!

File: dashboard/config/libraries/PoemBot.interpreted.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  var lineHeight = (400 - cursor) / poem.lines.length;
+  for (var i = 0; i < poem.lines.length; i++) {
+    var line = poem.lines[i];
+    var width = textWidth(line);
+    textSize(Math.min(defaultFontSize, defaultFontSize * (400 - outerMargin) / width));
+    text(line, 200, cursor);
+    textSize(defaultFontSize);
+    cursor += lineHeight;
```
Comment (breville) [inline dashboard/config/libraries/PoemBot.interpreted.js:37]: Are there any advantages to keeping this value an integer?  IIRC, some rendering paths in the browser end up with a performance and quality hit when rendered into a non-integer "subpixel" location, but I'm not sure if this is one of them.

File: dashboard/config/libraries/PoemBot.interpreted.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  if (poem.title) {
+    textSize(defaultFontSize * 2);
+    text(poem.title, 200, cursor);
+    textSize(defaultFontSize);
+    cursor += defaultLineHeight;
+  }
+  if (poem.author) {
+    cursor -= defaultLineHeight / 2;
+    textSize(16);
+    text("by: " + poem.author, 200, cursor);
```
Comment (breville) [inline dashboard/config/libraries/PoemBot.interpreted.js:25]: Do we need to think about non-en loc issues at some point?

File: dashboard/config/scripts/levels/New PoemBot Project.level [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
.../levelbuilder-studio.code.org/media?u=https%3A%2F%2Flevelbuilder-studio.code.org%2Fmedia%3Fu%3Dhttps%253A%252F%252Fstudio.code.org%252Fv3%252Fanimations%252F1alB-zQ8FPOdcMCQ8JPl2RP80uKiLDna7JAYAYRze7A%252Fa54f4f7d-6764-4345-abaa-b651abdca33e.png%253Fversion%253D6x.8TVua7c1V7jHF1sg5T3.LfPLdT7PU\",\r\n      \"frameSize\": {\r\n        \"x\": 280,\r\n        \"y\": 290\r\n      },\r\n      \"frameCount\": 1,\r\n      \"looping\": true,\r\n      \"frameDelay\": 12,\r\n      \"version\": \"6x.8TVua7c1V7jHF1sg5T3.LfPLdT7PU\"\r\n    },\r\n    \"64c242da-16c3-4261-9c57-936ce1f90aad\": {\r\n      \"name\": \"sick_mask\",\r\n      \"sourceUrl\": \"https://levelbuilder-studio.code.org/media?u=https%3A%2F%2Flevelbuilder-studio.code.org%2Fmedia%3Fu%3Dhttps%253A%252F%252Fstudio.code.org%252Fv3%252Fanimations%252F1alB-zQ8FPOdcMCQ8JPl2RP80uKiLDna7JAYAYRze7A%252F64c242da-16c3-4261-9c57-936ce1f90aad.png%253Fversion%253DYm9Z9V3PzU8puG_vRfdn2fsPcYtbd4.2\",\r\n      \"frameSize\": {\r\n        \"x\": 280,\r\n        \"y\": 290\r\n      },\r\n      \"frameCount\": 1,\r\n      \"looping\": true,\r\n      \"frameDelay\": 12,\r\n      \"version\": \"Ym9Z9V3PzU8puG_vRfdn2fsPcYtbd4.2\"\r\n    }\r\n  }\r\n}",
```
Comment (breville) [inline dashboard/config/scripts/levels/New PoemBot Project.level]: Should this file be trimmed at some point?

File: dashboard/config/scripts/levels/New PoemBot Project.level [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
...s\",\"toolbox_blocks\",\"block_pools\",\"preload_asset_list\",\"encrypted_examples\"],\"changed_by_id\":1,\"changed_by_email\":\"anjali+labs@code.org\"},{\"changed_at\":\"2021-06-21 08:56:39 -0400\",\"changed\":[\"start_blocks\",\"toolbox_blocks\",\"preload_asset_list\",\"encrypted_examples\"],\"changed_by_id\":1,\"changed_by_email\":\"anjali+labs@code.org\"},{\"changed_at\":\"2021-06-21 09:24:09 -0400\",\"changed\":[\"start_blocks\",\"toolbox_blocks\",\"preload_asset_list\",\"encrypted_examples\"],\"changed_by_id\":1,\"changed_by_email\":\"anjali+labs@code.org\"},{\"changed_at\":\"2021-06-21 10:35:35 -0400\",\"changed\":[\"start_blocks\",\"toolbox_blocks\",\"preload_asset_list\",\"encrypted_examples\"],\"changed_by_id\":1,\"changed_by_email\":\"anjali+labs@code.org\"},{\"changed_at\":\"2021-06-21 10:43:51 -0400\",\"changed\":[\"start_blocks\",\"toolbox_blocks\",\"preload_asset_list\",\"encrypted_examples\"],\"changed_by_id\":1,\"changed_by_email\":\"anjali+labs@code.org\"},{\"changed_at\":\"2021-06-21 13:28:41 -0400\",\"changed\":[\"start_blocks\",\"toolbox_blocks\",\"preload_asset_list\",\"encrypted_examples\"],\"changed_by_id\":1,\"changed_by_email\":\"anjali+labs@code.org\"}]",
```
Comment (breville) [inline dashboard/config/scripts/levels/New PoemBot Project.level]: Some of this history looks unrelated :)

**Review by breville — APPROVED** [review]
There is likely someone more qualified to review this, but just a few questions at first glance.  Looks like a nice start!

File: apps/src/sites/studio/pages/googleBlocklyWrapper.js [hunk 6]
```diff
@@ -245,9 +245,7 @@ function initializeBlocklyWrapper(blocklyInstance) {
   // so that our code generation logic still works with Google Blockly
   blocklyWrapper.Generator.blockSpaceToCode = function(name, opt_typeFilter) {
     const generator = blocklyWrapper.getGenerator();
-    if (!generator.isInitialized) {
-      generator.init(blocklyWrapper.mainBlockSpace);
-    }
+    generator.init(blocklyWrapper.mainBlockSpace);
```
Comment (maddiedierker) [inline apps/src/sites/studio/pages/googleBlocklyWrapper.js:248]: why don't we need to check if the generator has already been initialized?

File: dashboard/config/libraries/PoemBot.interpreted.js (hunk 3 above)
Comment (ajpal) [reply dashboard/config/libraries/PoemBot.interpreted.js:25]: Yes, FND has it on their backlog to help with i18n in July

File: dashboard/config/scripts/levels/New PoemBot Project.level (hunk 5 above)
Comment (ajpal) [reply dashboard/config/scripts/levels/New PoemBot Project.level]: I think this is related to the bug where you can't actually make a new level on levelbuilder. So every new level has to be cloned from something else, and I think a lot of this history carries over when you clone.

File: apps/src/sites/studio/pages/googleBlocklyWrapper.js (hunk 6 above)
Comment (ajpal) [reply apps/src/sites/studio/pages/googleBlocklyWrapper.js:248]: The Google Blockly generator clears its state each time you generate code, and expects that if you want to re-generate code you will reinitialize. This is different from our generator which maintained its state for the entire duration, and initializing was not idempotent so you only wanted to do it once at the beginning.

This didn't really matter for Flappy because there are no variables, but in Sprite Lab, the variable map in the generator was getting cleared after the first preview, and never reinitialized, so subsequent previews/runs weren't working.

File: dashboard/config/scripts/levels/New PoemBot Project.level (hunk 4 above)
Comment (ajpal) [reply dashboard/config/scripts/levels/New PoemBot Project.level]: I cleared out the `start_animations` because I know that's safe, but I don't want to edit this file directly since it's generated from LB
